hkmc2/shared/src/test/mlscript/nofib/minimax.mls

```diff
@@ -0,0 +1,387 @@
+  (0 :: 1 :: 0 :: Nil) ::
+  (0 :: 0 :: 1 :: Nil) :: Nil
+//│ > Cons {
+//│ >   head: Cons { head: 1, tail: Cons { head: 0, tail: [Cons] } },
+//│ >   tail: Cons {
+//│ >     head: Cons { head: 0, tail: [Cons] },
+//│ >     tail: Cons { head: [Cons], tail: [Nil] }
+//│ >   }
+//│ win7 = }
+
+let win8 =
+  (0 :: 0 :: 1 :: Nil) ::
+  (0 :: 1 :: 0 :: Nil) ::
+  (1 :: 0 :: 0 :: Nil) :: Nil
+//│ > Cons {
+//│ >   head: Cons { head: 0, tail: Cons { head: 0, tail: [Cons] } },
+//│ >   tail: Cons {
+//│ >     head: Cons { head: 0, tail: [Cons] },
+//│ >     tail: Cons { head: [Cons], tail: [Nil] }
+//│ >   }
+//│ win8 = }
+
+let wins = win1 :: win2 :: win3 :: win4 :: win5 :: win6 :: win7 :: win8 :: Nil
+//│ > Cons {
+//│ >   head: Cons {
+//│ >     head: Cons { head: 1, tail: [Cons] },
+//│ >     tail: Cons { head: [Cons], tail: [Cons] }
+//│ >   },
+//│ >   tail: Cons {
+//│ >     head: Cons { head: [Cons], tail: [Cons] },
+//│ >     tail: Cons { head: [Cons], tail: [Cons] }
+//│ >   }
+//│ wins = }
+
+fun showPiece(p) = if p is
+  X then nofibStringToList("X")
+  O then nofibStringToList("O")
+  Empty then nofibStringToList(" ")
+
+fun showRow(ps) = if ps is
+  p1 :: p2 :: p3 :: Nil then
+    append(append(showPiece(p1), nofibStringToList("|")), append(showPiece(p2), append(nofibStringToList("|"), showPiece(p3))))
+
+fun showBoard(rs) = if rs is
+  r1 :: r2 :: r3 :: Nil then
+    append(
+      showRow(r1),
+      append(
+        nofibStringToList("\n------\n"),
+        append(
+          showRow(r2),
+          append(
+            nofibStringToList("\n------\n"),
+            append(
+              showRow(r3),
+              nofibStringToList("\n\n")
+            )
+          )
+        )
+      )
+    )
+
+fun insert(p, ps, i) = if ps is Cons(p1, Cons(p2, Cons(p3, Nil))) and i ===
+  1 then p :: p2 :: p3 :: Nil
+  2 then p1 :: p :: p3 :: Nil
+  3 then p1 :: p2 :: p :: Nil
+
+fun empty_(x, r) = if
+  x === 1 and r is Empty :: _ :: _ :: Nil then true
+  x === 2 and r is _ :: Empty :: _ :: Nil then true
+  x === 3 and r is _ :: _ :: Empty :: Nil then true
+  else false
+
+fun empty(pos, board) = if board is r1 :: r2 :: r3 :: Nil and pos is
+  [1, x] then empty_(x, r1)
+  [2, x] then empty_(x, r2)
+  [3, x] then empty_(x, r3)
+
+fun placePiece(p, board, pos) = if
+  not(empty(pos, board)) then Nil
+  board is r1 :: r2 :: r3 :: Nil and pos is
+    [1, x] then (insert(p, r1, x) :: r2 :: r3 :: Nil) :: Nil
+    [2, x] then (r1 :: insert(p, r2, x) :: r3 :: Nil) :: Nil
+    [3, x] then (r1 :: r2 :: insert(p, r3, x) :: Nil) :: Nil
+
+fun fullBoard(b) = andd(map(
+  x => not(eqPiece(x, Empty)),
+  concat(b)
+))
+
+fun newPositions(piece, board) =
+  fun lscomp1(ls) = if ls is
+    Nil then Nil
+    x :: xs then
+      fun lscomp2(ls) = if ls is
+        Nil then lscomp1(xs)
+        y :: ys then [x, y] :: lscomp2(ys)
+      lscomp2(1 :: 2 :: 3 :: Nil)
+  concat(map(
+    pos => placePiece(piece, board, pos),
+    lscomp1(1 :: 2 :: 3 :: Nil)
+  ))
```

Comment:
```suggestion
  concat of map of
    pos => placePiece(piece, board, pos)
    lscomp1(1 :: 2 :: 3 :: Nil)
```

Reply:
Thanks! I have fixed this in the latest commit.